doc/archiveopteryx.man
.\" Copyright 2009 The Archiveopteryx Developers <info@aox.org>
.TH archiveopteryx 7 2014-03-10 aox.org "Archiveopteryx Documentation"
.SH NAME
Overview of Archiveopteryx.
.SH SYNOPSIS
.B /etc/init.d/archiveopteryx {start,stop}
.PP
.B $CONFIGDIR/*.conf
.SH DESCRIPTION
.nh
.PP
Archiveopteryx is a mail server system optimised for long-term storage
and heavy access. It comprises a set of server programs that provide
access to mail stored in a relational database (PostgreSQL).
.SH "GETTING STARTED"
The following five steps should suffice to install Archiveopteryx. If they
don't, or if anything is unclear, more detailed instructions are
available at http://archiveopteryx.org/installation
.SS "1. Install Archiveopteryx"
Usually, this is as simple as:
.IP
.B "apt-get install archiveopteryx"
(on Debian)
.br
.B "pkg install archiveopteryx"
(on FreeBSD)
.PP
http://archiveopteryx.org/installation explains what to do in
other cases.
.PP
Use
.IP
.B $LIBDIR/archiveopteryx start
.PP
to start Archiveopteryx using the default configuration from
.BR archiveopteryx.conf ,
which tries hard to be sensible.
.SS "2. Add a user"
Run
.IP
aox add user
.I name
.I password
.I address@domain
.PP
to create a username and password.
.SS "3. Inject some test mail"
You can use
.BR deliver (8)
to inject some mail you already have.
.PP
This command injects all the messages from a berkeley mbox file:
.IP
formail -s deliver yourname@domain < mbox
.PP
.SS "4. Check that you can read mail"
The mail you just delivered is in your inbox, and you can read it with
any IMAP or POP3 client.
.SS "5. Configure your MTA"
The normal way to deliver mail from your MTA into Archiveopteryx is via LMTP.
By default, Archiveopteryx listens on 127.0.0.1 port 2026.
.PP
On aox.org we describe how to configure some common MTAs to work with
Archiveopteryx. See e.g. http://archiveopteryx.org/postfix for
.BR postfix (1).
.SH ARCHITECTURE
.PP
Archiveopteryx consists of a number of frontend servers, each of which uses
backend servers:
.IP RDBMS
The RDBMS is where all the mail is stored. At present only PostgreSQL
is supported.
.IP logd
The log server is an internal server which filters log entries so
useful messages are logged while noise is suppressed.
.PP
Each server is described more fully in its own man page:
.BR logd (8),
.BR postgres (1)
and
.BR archiveopteryx (8)
for the frontend server, which serves IMAP, POP3, LMTP and/or SMTP and
perhaps webmail.
.PP
Normally, there is one
.BR logd (8)
process running. In addition there will be at least one
.BR postgres (1)
and one
.BR archiveopteryx (8)
process.
.PP
The servers use TCP to communicate internally, so they can be
distributed across a server cluster. Generally, one host will run the
database backend, one host (perhaps the same) will run
.BR logd (8)
and as many as required will run frontends.
.PP
The man pages for each server explain the use and configuration of
each. See
.BR archiveopteryx.conf (5)
for more about configuring Archiveopteryx in general.
.SH SECURITY
All Archiveopteryx servers run in chroot directories.
.PP
The user-facing servers run in a special jail directory. They have
neither read nor write access to this directory.
.PP
.BR logd (8)
runs chrooted to another directories. All of the servers close all open
files at startup and drop root privileges. By default they run as user
.IR $AOXUSER ,
group
.IR $AOXGROUP ,
although these names can be changed using
.BR archiveopteryx.conf (5).
.PP
The servers check that they lose prvileges as expected, and refuse to
start if they're too privileged.
.PP
Note that
.BR logd (8)
must have permission to create the
.IR logfile .
.SH MAIL STORAGE
Archiveopteryx does not store mail in the RFC-822 format. It parses each
message upon delivery, and stores a normalized representation, optimized
for fast and reliable search and categorization.
.PP
This offers the following advantages, among others:
.IP Performance
One example: Each address is stored exactly once, as Unicode and with
RFC-2047 encoding undone. Finding all messages sent from/to a given
name is extremely fast, because only a single SQL SELECT is necessary,
and it accesses just two small tables.
.IP "Long-term Stability"
Archiveopteryx handles today's common syntax problems and stores the
correct form in the database. Because of that, the mail reader which
looks at old mail in the year 2020 will not need to be bug compatible
with today's version of Microsoft Outlook.
.IP "Scalability and Flexibility"
Only the database size limits Archiveopteryx's capacity. Many other servers
limit individual folders to 2GB (or less), or cannot support more than
a few thousand subfolders/messages in a folder. With Archiveopteryx, you
don't need to invent workarounds for such artificial limitations.
.IP Robustness
Because mail is stored normalized and parsed, large attachments are
generally stored only once, and mail parsing exploits are rendered
harmless before reaching the MUA. A movie clip sent to a hundred
recipients is not a problem, because it's stored just once.
.SH LICENSING
Archiveopteryx is available under the PostgreSQL License.
.SH DEFAULTS
The configurable file and directory names in this build are as follows:
.IP SBINDIR
(where servers live) is
.IR $SBINDIR .
.IP BINDIR
(where other executables live) is
.IR $BINDIR .
.IP INITDIR
(where the startup script lives) is
.IR $INITDIR .
.IP MANDIR
(where manpages live) is
.IR $MANDIR .
.IP PIDFILEDIR
(where pidfiles live) is
.IR $PIDFILEDIR .
.IP LIBDIR
(where supporting files live) is
.IR $LIBDIR .
.IP JAILDIR
(the working directory of the user-facing servers) is
.IR $JAILDIR ,
and can be overridden using the
.I jail-directory
variable in
.BR archiveopteryx.conf (5).
.IP CONFIGDIR
(where the configuration files live) is
.IR $CONFIGDIR .
.IP LOGFILE
(the full name of the logfile) is
.IR $LOGFILE ,
and can be overridden using the
.I logfile
variable in
.BR archiveopteryx.conf (5).
.PP
These variables can be changed only by editing the file Jamsettings
and recompiling Archiveopteryx. Jamsettings also contains some variables
used only during compilation and/or installation, and some which
provide defaults for
.BR archiveopteryx.conf (5)
settings.
.SH FILES
.IP $CONFIGDIR/archiveopteryx.conf
contains the Archiveopteryx configuration.
.IP $LIBDIR/automatic-key.pem
contains a private key and self-signed certificate used by
.BR openssl (3).
.SH AUTHOR
The Archiveopteryx Developers, info@aox.org.
.SH VERSION
This man page covers Archiveopteryx version 3.2.0, released 2014-03-10,
http://archiveopteryx.org/3.2.0
.SH SEE ALSO
.BR archiveopteryx (8),
.BR archiveopteryx.conf (5),
.BR deliver (8),
.BR logd (8),
http://archiveopteryx.org


archiveopteryx/archiveopteryx.cpp
// Copyright 2009 The Archiveopteryx Developers <info@aox.org>

#include "scope.h"
#include "server.h"

#include "pop.h"
#include "imap.h"
#include "smtp.h"
#include "graph.h"

#include "tlsthread.h"
#include "flag.h"
#include "event.h"
#include "cache.h"
#include "mailbox.h"
#include "listener.h"
#include "database.h"
#include "dbsignal.h"
#include "selector.h"
#include "managesieve.h"
#include "spoolmanager.h"
#include "entropy.h"
#include "egd.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h> // exit()

#include <time.h> // time()


/*! \nodoc */


class StartupWatcher
    : public EventHandler
{
public:
    StartupWatcher(): EventHandler() {}
    void execute() {
        if ( Log::disastersYet() )
            ::exit( 1 );
        EventLoop::global()->setStartup( false );
    }
};


class ConnectionObliterator
    : public EventHandler
{
public:
    ConnectionObliterator()
        : EventHandler() {
        (void)new DatabaseSignal( "obliterated", this );
    }
    void execute() {
        List<Connection>::Iterator i( EventLoop::global()->connections() );
        while ( i ) {
            Connection * c = i;
            ++i;
            if ( !c->hasProperty( Connection::Listens ) &&
                 !c->hasProperty( Connection::Internal ) ) {
                Scope x( c->log() );
                log( "The database was obliterated" );
                c->close();
            }
        }
        EventLoop::freeMemorySoon();
        Cache::clearAllCaches( true );
    }
};


class ArchiveopteryxEventLoop
    : public EventLoop
{
public:
    ArchiveopteryxEventLoop()
        : fine( true ), unhappinessStarted( 0 ) {}
    ~ArchiveopteryxEventLoop();

    void freeMemory();

    bool fine;
    time_t unhappinessStarted;
};


ArchiveopteryxEventLoop::~ArchiveopteryxEventLoop()
{
}


void ArchiveopteryxEventLoop::freeMemory()
{
    EventLoop::freeMemory();
    uint fromOS = Allocator::allocatedFromOS();
    if ( fromOS < memoryUsage() ) {
        fine = true;
        return;
    }
    if ( fine ) {
        unhappinessStarted = ::time( 0 );
        fine = false;
    }

    // how long have we been in a faq-inducing state?
    if ( time( 0 ) < unhappinessStarted + 60 )
        return;

    // oh no. make the parent replace this process with another.
    if ( ::fork() > 0 )
        ::exit( 0 );
    // the process watcher will notice that the parent fork exited,
    // and start a replacement. in the child, we shut down fairly
    // quickly.
    stop( 60 );
}



int main( int argc, char *argv[] )
{
    Scope global;

    Server s( "archiveopteryx", argc, argv );
    EventLoop::setup( new ArchiveopteryxEventLoop );
    s.setup( Server::Report );

    bool security( Configuration::toggle( Configuration::Security ) );
    EString root( Configuration::text( Configuration::JailDir ) );

    if ( Configuration::toggle( Configuration::UseSmtp ) ||
         Configuration::toggle( Configuration::UseLmtp ) )
    {
        EString mc( Configuration::text( Configuration::MessageCopy ) );
        EString mcd( Configuration::text( Configuration::MessageCopyDir ) );
        if ( mc == "all" || mc == "errors" || mc == "delivered" ) {
            struct stat st;
            if ( mcd.isEmpty() )
                log( "message-copy-directory not set", Log::Disaster );
            else if ( ::stat( mcd.cstr(), &st ) < 0 || !S_ISDIR( st.st_mode ) )
                log( "Inaccessible message-copy-directory: " + mcd,
                     Log::Disaster );
            else if ( security && !mcd.startsWith( root ) )
                log( "message-copy-directory must be under jail directory " +
                     root, Log::Disaster );
        }
        else if ( mc == "none" ) {
            if ( Configuration::present( Configuration::MessageCopyDir ) )
                log( "Disregarding message-copy-directory (value " + mcd +
                     ") because message-copy is set to none " );
        }
        else {
            log( "Invalid value for message-copy: " + mc, Log::Disaster );
        }
    }


    EString sA( Configuration::text( Configuration::SmartHostAddress ) );
    uint sP( Configuration::scalar( Configuration::SmartHostPort ) );

    if ( Configuration::toggle( Configuration::UseSmtp ) &&
         Configuration::scalar( Configuration::SmtpPort ) == sP &&
         ( Configuration::text( Configuration::SmtpAddress ) == sA ||
           ( Configuration::text( Configuration::SmtpAddress ) == "" &&
             sA == "127.0.0.1" ) ) )
    {
        log( "smarthost-address/port are the same as smtp-address/port",
             Log::Disaster );
    }

    if ( Configuration::toggle( Configuration::UseLmtp ) &&
         Configuration::scalar( Configuration::LmtpPort ) == sP &&
         ( Configuration::text( Configuration::LmtpAddress ) == sA ||
           ( Configuration::text( Configuration::LmtpAddress ) == "" &&
             sA == "127.0.0.1" ) ) )
    {
        log( "smarthost-address/port are the same as lmtp-address/port",
             Log::Disaster );
    }

    if ( Configuration::toggle( Configuration::UseSmtpSubmit ) &&
         Configuration::scalar( Configuration::SmtpSubmitPort ) == sP &&
         ( Configuration::text( Configuration::SmtpSubmitAddress ) == sA ||
           ( Configuration::text( Configuration::SmtpSubmitAddress ) == "" &&
             sA == "127.0.0.1" ) ) )
    {
        log( "smarthost-address/port are the same as "
             "smtp-submit-address/port", Log::Disaster );
    }


    EString app =
        Configuration::text( Configuration::AllowPlaintextPasswords ).lower();
    if ( !( app == "always" || app == "never" ) )
        ::log( "Unknown value for allow-plaintext-passwords: " + app,
               Log::Disaster );
    if ( app == "never" &&
         Configuration::toggle( Configuration::UseTls ) == false &&
         Configuration::toggle( Configuration::AuthCramMd5 ) == false &&
         Configuration::toggle( Configuration::AuthDigestMd5 ) == false )
        ::log( "allow-plaintext-passwords is 'never' and use-tls is 'false', "
               "but only plaintext authentication mechanisms are allowed",
               Log::Disaster );

    EString apa =
        Configuration::text( Configuration::AllowPlaintextAccess ).lower();
    if ( !( apa == "always" || apa == "localhost" || apa == "never" ) )
        ::log( "Unknown value for allow-plaintext-access: " + apa,
               Log::Disaster );
    if ( apa == "never" &&
         Configuration::toggle( Configuration::UseTls ) == false )
        ::log( "allow-plaintext-access is 'never', but use-tls is 'false'",
               Log::Disaster );

    // set up an EGD server for openssl
    Entropy::setup();
    EString egd( root );
    if ( !egd.endsWith( "/" ) )
        egd.append( "/" );
    egd.append( "var/run/egd-pool" );
    (void)new Listener< EntropyProvider >( Endpoint( egd, 0 ), "EGD" );
    if ( !security ) {
        struct stat st;
        if ( stat( "/var/run/edg-pool", &st ) < 0 ) {
            log( "Security is disabled and /var/run/edg-pool does not exist. "
                 "Creating it just in case openssl wants to access it." );
            (void)new Listener< EntropyProvider >(
                Endpoint( "/var/run/edg-pool", 0 ), "EGD(/)" );
        }
    }
    if ( ::chmod( egd.cstr(), 0666 ) < 0 )
        log( "Could not grant r/w access to EGD socket", Log::Disaster );

    Listener< IMAP >::create(
        "IMAP", Configuration::toggle( Configuration::UseImap ),
        Configuration::ImapAddress, Configuration::ImapPort
    );
    Listener< IMAPS >::create(
        "IMAPS", Configuration::toggle( Configuration::UseImaps ),
        Configuration::ImapsAddress, Configuration::ImapsPort
    );
    Listener< POP >::create(
        "POP3", Configuration::toggle( Configuration::UsePop ),
        Configuration::PopAddress, Configuration::PopPort
    );
    Listener< POPS >::create(
        "POP3S", Configuration::toggle( Configuration::UsePops ),
        Configuration::PopsAddress, Configuration::PopsPort
    );
    Listener< ManageSieve >::create(
        "Sieve", Configuration::toggle( Configuration::UseSieve ),
        Configuration::ManageSieveAddress, Configuration::ManageSievePort
    );
    Listener< SMTP >::create(
        "SMTP", Configuration::toggle( Configuration::UseSmtp ),
        Configuration::SmtpAddress, Configuration::SmtpPort
    );
    Listener< LMTP >::create(
        "LMTP", Configuration::toggle( Configuration::UseLmtp ),
        Configuration::LmtpAddress, Configuration::LmtpPort
    );
    Listener< SMTPSubmit >::create(
        "SMTP-Submit", Configuration::toggle( Configuration::UseSmtpSubmit ),
        Configuration::SmtpSubmitAddress, Configuration::SmtpSubmitPort
    );
    Listener< SMTPS >::create(
        "SMTPS", Configuration::toggle( Configuration::UseSmtps ),
        Configuration::SmtpsAddress, Configuration::SmtpsPort
    );

    if ( Configuration::toggle( Configuration::UseTls ) ) {
        TlsThread::setup();
    }

    s.setup( Server::LogStartup );

    Listener< GraphDumper >::create(
        "Statistics", Configuration::toggle( Configuration::UseStatistics ),
        Configuration::StatisticsAddress, Configuration::StatisticsPort
    );

    EventLoop::global()->setMemoryUsage(
        1024 * 1024 * Configuration::scalar( Configuration::MemoryLimit ) );

    s.setup( Server::Finish );

    Database::setup();

    StartupWatcher * w = new StartupWatcher;

    Database::checkSchema( w );
    if ( security )
        Database::checkAccess( w );
    EventLoop::global()->setStartup( true );
    Mailbox::setup( w );

    SpoolManager::setup();
    Selector::setup();
    Flag::setup();
    IMAP::setup();

    if ( !security )
        (void)new ConnectionObliterator;

    s.run();
}
